### ISSUE TITLE
2FA: two factor authentification

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1133]
In GitLab by @orioltinto on Sep 19, 2023, 12:32

Hello @dbeltrankyl and @kinow,

### Summary

Unfortunately, one of the platforms that we are using implemented a 2FA to access the platform.
Do you have encountered this problem before?
Do you have a working solution?

Thanks.

()

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 12:42

Hello, 

The ECMWF platform uses 2FA, but it allows us to generate a 1-week certificate, and we use that. 

I can't have access to that platform, right? 

Could you show me the output? Is it similar to this one? https://stackoverflow.com/questions/53761170/ssh-with-2fa-in-python-using-paramiko

I can try to implement it and ask you to test it

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 19, 2023, 12:47

I guess it is something similar.

It requires the access password and afterwards a One Time Password that can be accessed through a phone or web app:

```term
password: 
!!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 
All LRZ cluster login nodes require now a second factor for authentication! 
Please refer to https://doku.lrz.de/x/EgDxBw 
Note for using push tokens: 
Without PIN directly press enter upon the 2FA prompt to receive the push token!
2FA: 
```

Maybe I can try to test with a simpler paramiko app if I can implement a solution.
Anything I should take into account?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 12:54

> Maybe I can try to test with a simpler paramiko app if I can implement a solution. Anything I should take into account?

Not really. Our current connect code can be easily extended. But.. if you can accomplish a non-interactive 2fa would be great otherwise:

Non-iterative downsides:

- nohup autosubmit run $expid & # won't work, as the 2FA will wait for a prompt? So the nohup part won't work.
- auto recovery that takes place whenever the remote platform is not available will be hung for the same reason as it will reconnect to the platform again.
- I'm not sure what will happen with the remote logs because they're inside a thread that will establish a new connection .. so the 2fa will be triggered again.

Edit: In the last one, my wondering is whether it will open infinite sockets that will crash soon or later autosubmit, or it just won't do anything.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 12:59

If non-iteractive version is not possible, I suggest to  contact them and ask them about the possibility of having some kind of certificate like ecmwf does

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 19, 2023, 13:10

I could connect with this snippet:

```python
import paramiko

HOST = "HOST_ADDRESS"
PORT = 22
USERNAME = 'USERNAME'


def interactive_auth_handler(title, instructions, prompt_list):
    answers = []
    for prompt, is_echo in prompt_list:
        answer = input(prompt)
        answers.append(answer)
    return answers


def main():
    ssh_client = paramiko.SSHClient()
    ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    transport = paramiko.Transport((HOST, PORT))
    try:
        transport.start_client()
        transport.auth_interactive(USERNAME, interactive_auth_handler)
        if transport.is_authenticated():
            ssh_client._transport = transport
            print("Authentication successful!")
            # Continue with your usual operations, for example:
            _, stdout, _ = ssh_client.exec_command('ls')
            print(stdout.read().decode('utf-8'))
        else:
            print("Authentication failed!")
            transport.close()
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()

```

Do you think that it is feasible to implement it in Autosubmit?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 14:09

Hello @orioltinto, 

I can implement it, but:

* Does it work with Public key + 2fa? If yes I can implement it
* This is an interactive version, so the downsides mentioned will raise

For context, we are using the high-level paramiko connect method to:

* Agent auth(ssh-agent)
* Proxycommand 
* Disable (only if incompatible) some algorithms ( disabled_algorithms={'pubkeys': ['rsa-sha2-256', 'rsa-sha2-512']} ) 

Basic example

```python
self._ssh.connect(self._host_config['hostname'], port, username=self.user, timeout=60, banner_timeout=60)
self.transport = self._ssh.get_transport()
self.transport.banner_timeout = 60
self._ftpChannel = paramiko.SFTPClient.from_transport(self.transport,window_size=pow(4, 12) ,max_packet_size=pow(4, 12) )
self._ftpChannel.get_channel().settimeout(120)
```

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 19, 2023, 15:44

Hi @dbeltrankyl,
I didn't manage to use the ssh key, also from the terminal it asks me for the password even having a ssh key.

The client initialization is a bit different. I didn't succeed initializing the client with the interactive handler using the **connect** method.


Another thing, as a workaround we were considering running autosubmit directly in the remote machine and use the LOCAL platform. However, we are not able to submit jobs from the login without ssh'ing again (which we can't because it asks again for a OTP). Is there a way of using the LOCAL platform as a slurm machine?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 19, 2023, 16:18

> Another thing, as a workaround we were considering running autosubmit directly in the remote machine and use the LOCAL platform. However, we are not able to submit jobs from the login without ssh'ing again (which we can't because it asks again for a OTP). Is there a way of using the LOCAL platform as a slurm machine?

We never tried to do that, but I don't think so. 

The local platform doesn't use paramiko but has no method to communicate with the scheduler. It acts similarly to PS platform. 

To do that, I would need to implement a "slurm-local" type that doesn't exist now. Even if you define 

```
PLATFORMS:
  LOCAL:
    TYPE: slurm
    HOST: localhost
```

It will use paramiko to connect to the host-specified

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 19, 2023, 16:30

If it relies on paramiko it is not going to work since it is going to ask for credentials.

I was taking a look at the code. Would that be possible to subclass `SlurmPlatform` and simply overwrite the methods `send_file` and `send_command`? Maybe I'm missing important points here but at first glance it doesn't look like a lot of work.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 20, 2023, 09:05

Hello, 

I think it is not only that, 

I need to update many variables to point to local instead of remote..

At first glance, I see possible modifications in 

send_command, send_file, get_file, move_file, remote_permissions, submitter file, and maybe also the submit_ready_jobs function, check wrapper files, and others I have to see. 

Later, I will take a more in-depth look

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 20, 2023, 15:57

Hello @orioltinto,

Could you clone https://earth.bsc.es/gitlab/es/autosubmit/-/tags/v2fa_test and test it?  I don't want to upload an untested code to pip

To activate the 2fa, you only have to put 

```
PLATFORMS:
 PLATFORM:
   ...
   2FA: True
```

About the pubkey not allowing in the 2FA .. the issue is that the password is not protected and will be stored in the log as plain text. It is another downside and a dangerous one.

I also worked a bit on implementing a local slurm ( as a new type), but I'll need more time as I need to perform other tasks

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 20, 2023, 16:50

**EDIT**: It was a mistake on my end, I added the _2FA: True_ parameter to the wrong place.

Not working right now.

On **autosubmit run a000** I got:

```
[local] Correct user privileges for host localhost
[local] Connection successful to host localhost
Traceback (most recent call last):
  File "/home/oriol/autosubmit/autosubmit_src/autosubmit/autosubmit.py", line 2045, in run_experiment
    job_list, submitter , exp_history, host , as_conf, platforms_to_test, packages_persistence, _ = Autosubmit.prepare_run(expid, notransitive,start_time, start_after, run_only_members)
  File "/home/oriol/autosubmit/autosubmit_src/autosubmit/autosubmit.py", line 1985, in prepare_run
    Autosubmit.restore_platforms(platforms_to_test)
  File "/home/oriol/autosubmit/autosubmit_src/autosubmit/autosubmit.py", line 2331, in restore_platforms
    raise AutosubmitCritical("Issues while checking the connectivity of platforms.", 7010, issues + "\n" + ssh_config_issues)
log.log.AutosubmitCritical:  

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/oriol/autosubmit/autosubmit_src/bin/autosubmit", line 37, in main
    Autosubmit.parse_args()
  File "/home/oriol/autosubmit/autosubmit_src/autosubmit/autosubmit.py", line 658, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive,args.start_time,args.start_after, args.run_only_members, args.profile)
  File "/home/oriol/autosubmit/autosubmit_src/autosubmit/autosubmit.py", line 2048, in run_experiment
    raise AutosubmitCritical(e.message, 7014, e.trace)
log.log.AutosubmitCritical:  
[ERROR] Trace: 
[LRZ] has configuration issues.
 Check that the connection is passwd-less.(ssh ra57nun@LRZ-kcs)
 Check the parameters that build the root_path are correct:{scratch_dir/project/user} = {/gpfs/scratch/uh211/ra57nun/lxcusers/ra57nun}
Timeout connection this is an PARAMIKO SSHEXCEPTION: indicates that there is something incompatible in the ssh_config for host:LRZ-kcs
 maybe you need to contact your sysadmin
 [CRITICAL] Issues while checking the connectivity of platforms. HINT: check the CUSTOM_DIRECTIVE syntax in your jobs configuration files. [eCode=7014]

```

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 20, 2023, 16:59

Wait, it was a mistake on my side. I managed to connect and get some jobs properly executed.
Minor thing: the first time it asks for the password and OTP the user input is completely visible.

Then, it asked me for password and OTP for another 3 times. Not sure if it was because I made a mistake with the passwords or not.

It looks like working, thank you!

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 21, 2023, 08:47

> Minor thing: the first time it asks for the password and OTP the user input is completely visible.

Yes, I noticed that too, not sure how to fix it. I tried with getpass api but I can still see what I'm typing*. According to internet, it seems that depending on the platform and terminal it may or not may work and it never will work in a shell.

Alteast it seems that it is not being stored in the ASLOGS/run.log

At the end, the best thing would be to use a pkey + interactive 2fa but I can't find how in the paramiko api if it is even possible.

> Then, it asked me for password and OTP for another 3 times. Not sure if it was because I made a mistake with the passwords or not.

Strange, the thing is that I don't have access to 2FA platform so I can't test that. 


I can merge it like it is now, adding some warnings and security concerns messages if it is good enough for you

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 21, 2023, 10:14

Hi @dbeltrankyl,

I'm familiar with this problem with getpass, from Pycharm it doesn't work as expected.
However, I believe that in most terminals it works well, so I guess that is best to use this solution even if does not work for all cases.

I tried it myself in **interactive_auth_handler** and it works well.

I still have a slight problem: the first time that the client starts autosubmit is still not logging the status of the jobs and one can login without problem. But the second time that it tries to authenticate (I guess that this is related to retrieving the remote logs) it mixes the password prompts with the logs and it is quite confusing. Related with this, we have some users that used to launch autosubmit from a jupyter notebook, and in this case also the frequent autosubmit logs are a bit unconvinient. Is there a way of configuring autosubmit to get log updates only when there's a change in the status?

Other than that I think that the change is good to go.

Thanks again.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 21, 2023, 10:53

> I still have a slight problem: the first time that the client starts autosubmit is still not logging the status of the jobs and one can login without problem. But the second time that it tries to authenticate (I guess that this is related to retrieving the remote logs) it mixes the password prompts with the logs and it is quite confusing. Related with this, we have some users that used to launch autosubmit from a jupyter notebook, and in this case also the frequent autosubmit logs are a bit unconvinient. Is there a way of configuring autosubmit to get log updates only when there's a change in the status?

Could you try to put 

```
PLATFORMS:
  PLATFORM:
    DISABLE_RECOVERY_THREADS: "True"
```

Notice that it is not a boolean as this is untouched from AS 3

This is an old function that uses an old code to avoid spawning threads while recovering the logs. 

It won't reconnect to the platform, it will slowdown Autosubmit and won't recover the logs for vertical wrappers. If it works for you I'll adapt in another branch the function to have all the functionalities of the normal recovery

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 21, 2023, 11:24

It looks like working but it is still asking me for the password+OTP 4 times.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 21, 2023, 11:26

Only at the beginning or during the execution? Could you send me the tmp/ASLOGS logs? ( check that the password is indeed not stored first )

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 21, 2023, 11:37

Yes, at the beginning of the **autosubmit run**. As I was saying before, after the first login it is a bit confusing as well because the autosubmit outputs start to overlap with the password requests.

I don't see any password nor anything related with connecting to the platform in the run log:
[20230921_112153_run.log](https://earth.bsc.es/gitlab/es/autosubmit/uploads/60e9115858f8738fda320b9ea60e5fe6/20230921_112153_run.log)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 21, 2023, 15:56

Hello @orioltinto , 

Could you test v2fa_test2 https://earth.bsc.es/gitlab/es/autosubmit/-/tags/v2fa_test2 ? 

Now it asks for a password and OTP only once across all autosubmit run ( hopefully ).

Then, each time it reconnects, it will reuse the same values and try again. If it doesn't work, it will just  raise an exception. ( With this I mean that autosubmit will work until there is an issue with the remote platform ) 

In each reconnection, there is a "press enter to continue" to give time to the user to accept the push notification. There is also a timeout mechanism to avoid touching autosubmit in future reconnections. Can be set with `2FA_TIMEOUT` parameter.

```
PLATFORMS:
  PLATFORM:
    2FA: True
    2FA_TIMEOUT: 10 # default 300 seconds
```

When using autosubmit, We can try with a push otp. 

This otp only requires an empty prompt after the user clicks on the "yes" in that notification ( or that I understood from the hpc docs ). So Autosubmit could work as long as that "yes" is clicked and the user press enter in the autosubmit console or the 2FA_timeout is reach.


Other OTP methods that require a pin or a token will work only once. So the recovery mechanism won't work. ( Unless it has a lifetime higher than 1 ) 

With the OTP push method, the recovery_threads to true should work with the adaptations made... but the user will get a notification for each job log retrieval. So you would want to keep it to disabled=True until #554 is solved

I'm not sure if this method will work, as I don't know what is inside the prompt_list, basically the change is that instead of having to prompt the password + otp in realtime, it will be prefilled

> 	PUSH token: Your ssh login to the Linux Cluster triggers a notification sent to your mobile device. By accepting the notification the second factor will be sent back to the LRZ server and the login procedure completes automatically.

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 21, 2023, 16:38

Hi @dbeltrankyl ,
I think that by design the OTP is not going to grant you access more than one time, even if the second time happens right after the other. I believe that the approach of storing the OTP can't work.

I tried to login myself twice with the same OTP and couldn't do it even if it was within the valid time of the password.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 22, 2023, 08:52

Hello @orioltinto ,

Yes, I thought so, but if you use the PUSH otp,

* It will send you a notification on the web/mobile device
* It requires an empty prompt in the terminal, 

So on the Autosubmit side, it should work if the user presses the push notification before the 2FA_timeout, no?

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 22, 2023, 11:57

Hi @dbeltrankyl ,
so far I only used TOTPs.
I'm trying to get the PUSH otp's working but so far I'm not successful.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 26, 2023, 15:48

Hello @orioltinto , 

I've added the tag v4.0.93 ( not yet uploaded to anything) 

Could you confirm that OTP is working on that version? 

There are three parameters for OTP now:

* 2FA: Allows to enable 2FA. Default: false
* 2FA_METHOD: Allows to set the OTP method, default: token possible values token/push
* 2FA_TIMEOUT: Allows to set a timeout when the push method is used. Default 300 seconds

All 2FA methods: will always ask for a password one time per platform at the beginning. 

* METHOD push: Won't ask for otp code, just for prompt enter or wait the timeout when the user click on the mobile device notification. This should be the ideal method. 
* METHOD token/others: Will ask for otp one time per connection. May cause issues while recovering log / or remote platform issues

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 26, 2023, 16:10

Hi @dbeltrankyl ,
I was working on this.

Your implementation has some minor bugs. Here's a version I got working:

```python
    def interactive_auth_handler(self, title, instructions, prompt_list):
        answers = []
        # Walk the list of prompts that the server sent that we need to answer
        for prompt_, _ in prompt_list:
            prompt = str(prompt_).strip().lower()
            # str() used to to make sure that we're dealing with a string rather than a unicode string
            # strip() used to get rid of any padding spaces sent by the server
            if "password" in prompt:
                answers.append(self.pw)
            elif "token" in prompt or "2fa" in prompt or "otp" in prompt:
                if self.two_factor_method == "push":
                    answers.append("")
                elif self.two_factor_method == "token":
                    twofactor_nonpush = input("Please type the 2FA/OTP/token code: ")
                    answers.append(twofactor_nonpush)
        return tuple(answers)
```

Some things:
- I don't understand the purpose of the **inputimeout** thing.
- The elif "token" ... had the logic reversed.
- I could test the push token that works well and the TOTP token that works but it's a bit confusing with the other autosubmit output in the middle.

Thanks for dealing with this. Let me know if you want me to test something else.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 27, 2023, 08:49

Thanks Oriol. I'll apply the fix. 

> * I don't understand the purpose of the **inputimeout** thing.

This is for the push method. There are two proposes both are for running time:

* Whenever the remote logs are retrieved, the platform will make a new connection that may ask for a password and otp. The idea is to avoid the necessity of pressing enter in the terminal where Autosubmit is launched and instead only click on the PUSH notification.

* The remote platform may lose the connection. If this is the case, Autosubmit will try to reconnect to all platform affected. This process will also wait for a OTP. Again, the user could click on the notification, and autosubmit will continue. 

Ideally, the user would only have to prompt the password at the start of the autosubmit run and then just click on the push notifications.

> * The elif "token" ... had the logic reversed.

oops

> I could test the push token that works well and the TOTP token that works but it's a bit confusing with the other autosubmit output in the middle.

Which output do you mean?

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 27, 2023, 09:05

What I mean is that autosubmit keeps printing the usual log (i.e. 2 of 2 jobs remaining (08:59)) while asking for the OTP again.


```log
Autosubmit is running with v4.0.93
Checking the connection to all platforms in use
Please type the 2FA/OTP/token code: 
[LRZ] Correct user privileges for host LRZ
[LRZ] Connection successful to host LRZ
[local] Correct user privileges for host localhost
[local] Connection successful to host localhost


2 of 2 jobs remaining (08:59)
No jobs to check for platform LRZ
No jobs to check for platform local
a000_TRANSFER_PROJECT submitted
Call from a000_TRANSFER_PROJECT with status SUBMITTED
(ra57nun@kcs-login.cos.lrz.de) Password: 

2 of 2 jobs remaining (08:59)
No jobs to check for platform LRZ
Checking 1 jobs for platform local
Job a000_TRANSFER_PROJECT is RUNNING

(ra57nun@kcs-login.cos.lrz.de) !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 2FA !!! 
All LRZ cluster login nodes require now a second factor for authentication! 
Please refer to https://doku.lrz.de/x/EgDxBw 
Note for using push tokens: 
Without PIN directly press enter upon the 2FA prompt to receive the push token!
2FA: 
(ra57nun@kcs-login.cos.lrz.de) Password: 

2 of 2 jobs remaining (08:59)
No jobs to check for platform LRZ
Checking 1 jobs for platform local
Job a000_TRANSFER_PROJECT is RUNNING


2 of 2 jobs remaining (09:00)
No jobs to check for platform LRZ
Checking 1 jobs for platform local
Job a000_TRANSFER_PROJECT is RUNNING

```

Another minor thing. With this implementation autosubmit asks for the password even with commands that don't need it, like **autosubmit create**. Maybe the password could be saved the first time that is needed.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 27, 2023, 10:29

> while asking for the OTP again.

About the password, it should be already stored, I've moved the prompt to the connect function and it should ask it only once. 

About the OTP, it will ask as many times as jobs completed.. that's because, in each log retrieval, there is a new connection taking place. 

We can use the ( with the downgrade on the speed ) 

```yaml 
PLATFORMS:
  PLATFORM:
    DISABLE_RECOVERY_THREADS: "True"
```

Or use the 2FA_METHOD = push and define a low timeout. 

Could you try again? I've https://earth.bsc.es/gitlab/es/autosubmit/-/tags/v4.0.93 reupload this ( as it was not uploaded to anywhere)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 27, 2023, 10:29

> Another minor thing. With this implementation autosubmit asks for the password even with commands that don't need it, like **autosubmit create**. Maybe the password could be saved the first time that is needed.

This should be fixed aswell

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 27, 2023, 12:27

Oh in fact I see that the second time it was asking me for the password and OTP it was due to the transfer job that as trying to ssh to the remote machine. I'll have to find a way to deal with that as well.

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 28, 2023, 15:18

Hi @dbeltrankyl , thank you for your time today. It was very useful.

Any timeline to have this in the pypi version so I can talk with my colleagues about using it?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Sep 28, 2023, 15:55

Hello @orioltinto, 


I've just uploaded it into pypi ~~4.0.94~~ 4.0.95. 

This version doesn't have the input timeout stuff

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 28, 2023, 16:51

Hi @dbeltrankyl, great thanks.

Maybe inside the interactive auth method you could add an informative message saying that the PUSH notification was sent. It might be helpful for a distracted user.

**Comment by kinow** [comment]
In GitLab by @orioltinto on Sep 29, 2023, 10:35

My colleague in LMU mentioned that it is easy to setup a 2FA option in a server using the google authenticator:

https://github.com/google/google-authenticator-libpam

I share it here in case it can be useful for you to set up something to test this 2FA thing.

**Comment by kinow** [comment]
Looks interesting! Maybe we could check if there are any Docker images that already provide something similar, e.g. https://hub.docker.com/r/saintd/2fa/dockerfile.

If so we could try to create a test setup (and someday even use in some functional test -- which alas is not possible now as we don't support Docker runnners at BSC's GitLab).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 5, 2023, 08:27

Added the notification, it will be present on next release.

I'll close the issue since the 2FA is working, please reopen it if you find any issue

**Comment by kinow** [comment]
In GitLab by @orioltinto on Oct 5, 2023, 15:27

Hi again @dbeltrankyl,

unfortunately a user got banned because of many failed authentications.
I didn't tell them to use the **DISABLE_RECOVERY_THREADS** option and maybe that is enough to prevent the failed authentications. I reopened the issue just to be sure.

Some info from the user:

* Q: Were the successful logins that were problematic or only the unattended ones?
* A: They said currently no daily limit on the number of the successful logins. Probably daily counts of unattended login was the problem. They did not tell me the number of allowed mistakes.
* Q: Do you know how many login attempts were happening?
* A: In my case every job requested one 2FA login. So in total about 120 logins for each of the experiment dates.
* Q: Do you know if it was only a single workstation that was banned?
* A: Yes. Precisely speaking a single IP address. After the ban, I was still able to access from another server e.g. jupyterlab terminal. However once I ran the autoicon there, that server also got banned.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 5, 2023, 15:43

Hello @orioltinto ,

Did he was using the push notification or the token? 

> * A: In my case every job requested one 2FA login. So in total about 120 logins for each of the experiment dates.

With DISABLE_RECOVERY_THREADS this shouldn't happen, as it won't generate a new connection

**Comment by kinow** [comment]
In GitLab by @orioltinto on Oct 5, 2023, 15:45

It was using push tokens.

Great, then it shouldn't happen again, thanks.
Does it make sense to disable these threads by default when using 2FA?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 5, 2023, 15:49

I think so. Tell me if you find issues while using it

Additionally, if it delays the workflow too much, we could consider adding a parameter to disable the logs recovery. The users would have to go into the remote hpcarch to find them.

Maybe we can also add a flag to `autosubmit recovery` to recover all logs without touching the status of the jobs. So the users would only have to run it after the run finishes

**Comment by kinow** [comment]
In GitLab by @orioltinto on Oct 6, 2023, 11:40

It seems that the **DISABLE_RECOVERY_THREADS** prevented the connection issue, thanks. I would suggest to set it True by default when using 2FA as it looks like it is a problem that others might encounter in the future as well.
